### PR TITLE
Xarray <-> Kerchunk backend

### PR DESCRIPTION
**Pull request by norlandrhagen**
Based off this [issue](https://github.com/fsspec/kerchunk/issues/360#issuecomment-1751389780) opened by @TomAugspurger and the [Pangeo Discourse post](https://discourse.pangeo.io/t/making-kerchunk-as-simple-as-a-toggle/3690) opened by @ahuang11.

```
import xarray as xr
import kerchunk

ds = xr.open_dataset(<path_to_ref_file>, engine="kerchunk")


```

### CODE REVIEW

File: kerchunk/xarray_backend.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            _, ext = os.path.splitext(filename_or_obj)
+        except TypeError:
+            return False
+        return ext in {".json", ".parquet"}
+
+    description = "Open Kerchunk References with Xarray"
+    url = "https://fsspec.github.io/kerchunk/"
+
+
+def my_open_dataset(filename_or_obj, **kwargs):
```
Comment (TomAugspurger) [inline kerchunk/xarray_backend.py:28]: Probably want this to have a meaningful name.

And I suspect we'll want to forward keyword arguments to two places, so we'll need `reference_fs_options` (or `storage_options`?) and `open_dataset_options` instead of `**kwargs`.

**Review by TomAugspurger — COMMENTED** [review]
Thanks. It'd be good to add a basic test too.

File: kerchunk/xarray_backend.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import os
+from fsspec.implementations.reference import ReferenceFileSystem
+
+
+class KerchunkBackend(BackendEntrypoint):
+    def open_dataset(
+        self,
+        filename_or_obj,
+        **kwargs,
+    ):
```
Comment (ahuang11) [inline kerchunk/xarray_backend.py:12]: Should there be a docstring?

File: kerchunk/xarray_backend.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        except TypeError:
+            return False
+        return ext in {".json", ".parquet"}
+
+    description = "Open Kerchunk References with Xarray"
+    url = "https://fsspec.github.io/kerchunk/"
+
+
+def my_open_dataset(filename_or_obj, **kwargs):
+    fs = ReferenceFileSystem(fo=filename_or_obj, **kwargs)
```
Comment (ahuang11) [inline kerchunk/xarray_backend.py:29]: Should kwargs go here or `open_dataset`?

**Review by ahuang11 — COMMENTED** [review]
Awesome!! Thanks for creating this; I think it'll be immensely helpful

**Comment by norlandrhagen** [comment]
Np! Feel free to add, suggest or give it a try. Credit to @TomAugspurger for outlining it all.

File: kerchunk/xarray_backend.py (hunk 1 above)
Comment (keewis) [reply kerchunk/xarray_backend.py:28]: I'd probably have a signature of
```python
def open_reference_dataset(filename_or_obj, storage_options, open_dataset_kwargs):
    pass
```
(note, no `**kwargs`, since those will be caught by the backend entrypoint's `open_dataset`)

If you go with `storage_options`, you might have to disallow passing `fo` in there. If not, you could make `target_protocol`, `target_options`, `remote_protocol` and `remote_options` parameters (or combine them somehow into `target_options` and `remote_options`, not sure). Whatever you decide, `open_dataset_parameters` should list all the options for the backend.

File: kerchunk/xarray_backend.py (hunk 2 above)
Comment (keewis) [reply kerchunk/xarray_backend.py:12]: We currently don't have guidelines on how to best document entrypoints. It would be good to figure those out (and maybe standardize in a sphinx extension), but for now you will most likely only see the docstring when reading the code or when inspecting the entrypoint before directly passing it to `xr.open_dataset`

File: setup.py [hunk 4]
```diff
@@ -24,6 +24,7 @@
         "netcdf3": ["scipy"],
     },
     entry_points={
+        "xarray.backends": ["kerchunk=kerchunk.xarray_backend:KerchunkBackend"],
```
Comment (keewis) [inline setup.py:27]: no need to add the entrypoint in both `setup.cfg` and `setup.py` (or rather, it's probably better not to do that).

As a side note, I'd also recommend transitioning to `pyproject.toml` since that's the most stable way going forward (but that would definitely have to be a separate PR).

**Comment by norlandrhagen** [comment]
Thanks for all the feedback @martindurant, @TomAugspurger , @ahuang11 and @keewis!

File: setup.py (hunk 4 above)
Comment (martindurant) [reply setup.py:27]: Would be happy to see a PR.

**Review by TomAugspurger — APPROVED** [review]

**Comment by martindurant** [comment]
Let's give it a spin!
